### PR TITLE
Adding support for IIS profiling events in KUDU's process explorer

### PR DESCRIPTION
**Pull request by puneet-gupta**
This code will create a checkbox called "Collect IIS events" next to the START PROFILING button in process explorer in KUDU and this will allow to collect IIS ETW events, along with thread time events for Perfview and they will help in troubleshooting slow performance issues in Azure App Service

### CODE REVIEW

**Comment by safihamid** [comment]
Please combine all the commits into a single commit for the PRs.

File: Kudu.Services.Web/Content/Scripts/ProcessExplorer.js [hunk 1]
```diff
@@ -370,17 +393,20 @@ var Process = (function () {
         }, false);
         $(profilingButton).css("width", "138px");
 
-        if (this._json.is_profile_running) { // highlight button if the profiler has started
```
Comment (safihamid) [inline Kudu.Services.Web/Content/Scripts/ProcessExplorer.js]: why is this removed?

**Comment by davidebbo** [comment]
@safihamid having multiple commits should not affect your ability to review, since it shows you the overall change to comment on. Only when it comes time to merge it in do we need to worry about squashing.

File: Kudu.Services/Diagnostics/ProfileManager.cs [hunk 2]
```diff
@@ -18,16 +18,24 @@ internal static class ProfileManager
     {
         private const int ProcessExitTimeoutInSeconds = 180;
 
+        private const string SERVICE_PROFILER_AGENT_GUID = "F5091AA9-80DC-49FF-A7CF-BD1103FE149D";
```
Comment (safihamid) [inline Kudu.Services/Diagnostics/ProfileManager.cs]: Please rename all const names to Pascal casing like the rest of Kudu files. like ServiceProfilerAgentGuid

File: Kudu.Services/Diagnostics/ProfileManager.cs [hunk 3]
```diff
@@ -39,34 +47,96 @@ internal static class ProfileManager
...
+                EnsureIdleTimer(iisProfiling);
 
                 return new ProfileResultInfo(HttpStatusCode.OK, string.Empty);
             }
         }
 
-        internal static async Task<ProfileResultInfo> StopProfileAsync(int processId, ITracer tracer = null)
+        private static async Task<ProfileResultInfo> StartIISSessionAsync(int processId, int profilingSessionId, ITracer tracer = null)
+        {
+            string arguments = System.Environment.ExpandEnvironmentVariables(string.Format("start {0} /scratchLocation:\"%LOCAL_EXPANDED%\\Temp\" /loadAgent:{1};ServiceProfilerAgent.dll /monitor", profilingSessionId, SERVICE_PROFILER_AGENT_GUID));
```
Comment (safihamid) [inline Kudu.Services/Diagnostics/ProfileManager.cs]: Are you sure you will need the /monitor switch? why is it needed? we didn't need it for CPU profiling. please look at the help for VSDiagnostics.exe

File: Kudu.Services/Diagnostics/ProfileManager.cs [hunk 4]
```diff
@@ -39,34 +47,96 @@ internal static class ProfileManager
...
+            var command = Task.Run(async () =>
+            {
+                await ExecuteProfilingCommandAsync(arguments, tracer);
+            });
+
+            // implementing a retry logic because it can so happen
+            // that the intial profiler command may take some time to start
+            // so max we will wait till 9 seconds before giving up
+            int retryCount = 0;
```
Comment (safihamid) [inline Kudu.Services/Diagnostics/ProfileManager.cs]: There should really be no retry logic here. Tasks can Wait on completion of a task or ContinueWith when a task finishes.

File: Kudu.Services/Diagnostics/ProfileManager.cs [hunk 5]
```diff
@@ -39,34 +47,96 @@ internal static class ProfileManager
...
+
+            ProfileResultInfo profileProcessResponse = new ProfileResultInfo(HttpStatusCode.InternalServerError, "Trying to initialize profiler");
+
+            while (retryCount < 2 && profileProcessResponse.StatusCode != HttpStatusCode.OK)
+            {
+                await Task.Delay(3000);
```
Comment (safihamid) [inline Kudu.Services/Diagnostics/ProfileManager.cs]: This Delay is always happening regardless which is redundant.

File: Kudu.Services/Diagnostics/ProfileManager.cs [hunk 6]
```diff
@@ -39,34 +47,96 @@ internal static class ProfileManager
...
+                    if (profileProcessResponse.StatusCode != HttpStatusCode.OK)
+                    {
+                        return profileProcessResponse;
+                    }
                 }
 
                 // This may fail if we got 2 requests at the same time to start a profiling session
                 // in that case, only 1 will be added and the other one will be stopped.
-                if (!_profilingList.TryAdd(processId, new ProfileInfo(profilingSessionId)))
+                if (!_profilingList.TryAdd(processId, new ProfileInfo(profilingSessionId,iisProfiling)))
```
Comment (safihamid) [inline Kudu.Services/Diagnostics/ProfileManager.cs]: nit: space after comma.

File: Kudu.Services/Diagnostics/ProfileManager.cs [hunk 7]
```diff
@@ -18,16 +18,24 @@ internal static class ProfileManager
...
+        private const string IIS_WWWSERVER_GUID = "3A2A4E84-4C21-4981-AE10-3FDA0D9B0F83";
+        private const string DIAGNOSTICS_HUB_AGENT_GUID = "4EA90761-2248-496C-B854-3C0399A591A4";
+
         private static ConcurrentDictionary<int, ProfileInfo> _profilingList = new ConcurrentDictionary<int, ProfileInfo>();
         private static object _lockObject = new object();
 
         // The profiling session timeout, this is temp fix before VS2015 Update 1.
         private static readonly TimeSpan _profilingTimeout = TimeSpan.FromMinutes(15);
+
+        private static readonly TimeSpan _profilingIISTimeout = TimeSpan.FromMinutes(3);
```
Comment (safihamid) [inline Kudu.Services/Diagnostics/ProfileManager.cs]: Can you deploy a standard asp.net site or web api, on a small dedicated worker, enable IIS profiling and load test it with WCAT or some other http stress testing tool for 3 minutes and share the file? Also plz log the timestamp, I want to see how big the file would be and how the VM behaved during that time.
If the file size is too big that even PerfView has problem opening it or if the VM is having Disk Space issues on D drive or the site's temp folder then we must lower the timeout value

File: Kudu.Services/Diagnostics/ProfileManager.cs (hunk 5 above)
Comment (puneet-gupta) [inline Kudu.Services/Diagnostics/ProfileManager.cs]: Just saw your first message and spent hours figuring out how to squash multiple commits in to 1. got that done. Will work on the rest of stuff by tomorrow

Comment (puneet-gupta) [inline Kudu.Services/Diagnostics/ProfileManager.cs]: Just saw your first message and spent hours figuring out how to squash multiple commits in to 1. got that done. Will work on the rest of stuff by tomorrow

File: Kudu.Services/Diagnostics/ProfileManager.cs (hunk 3 above)
Comment (pharring) [inline Kudu.Services/Diagnostics/ProfileManager.cs]: You can remove /monitor (it's useful when running from the command line for testing, but not in production)

**Comment by puneet-gupta** [comment]
Addressed following changes in latest commit

1. Changed contant names to pascal naming convention
2. Fixed space before iistracing for the comment 'nit: space after comma'
3. Removed monitor switch from profiler
4. Removed all unwanted logic for retry etc. which I had put earlier for the monitor switch which wasnt necessary.
5. Put back the line in processexplorer.js which got removed accidently

File: Kudu.Services/Diagnostics/ProfileManager.cs [hunk 8]
```diff
@@ -18,16 +18,24 @@ internal static class ProfileManager
     {
         private const int ProcessExitTimeoutInSeconds = 180;
 
+        private const string ServiceProfilerAgentGuid = "F5091AA9-80DC-49FF-A7CF-BD1103FE149D";
```
Comment (pharring) [inline Kudu.Services/Diagnostics/ProfileManager.cs]: Let's call this ```UserModeCustomProviderAgentGuid```. (While it is bundled in the Service Profiler DLL, that's just for convenience.)
Once you've done that rename, then it frees up the name ```ServiceProfilerAgentGuid``` for what is currently ```ServiceProfilerProviderGuid``` (1C92B...)

File: Kudu.Services/Diagnostics/ProfileManager.cs [hunk 9]
```diff
@@ -18,16 +18,24 @@ internal static class ProfileManager
...
+        private const string ServiceProfilerProviderGuid = "1C92BA2A-A990-480F-A02F-40068871CAAC";
+        private const string IISWWWServerGuid = "3A2A4E84-4C21-4981-AE10-3FDA0D9B0F83";
```
Comment (pharring) [inline Kudu.Services/Diagnostics/ProfileManager.cs]: I would call this something with the word "Provider" in it. e.g. ```IISWWWServerProviderGuid``` (it's a GUID for an ETW provider)

File: Kudu.Services/Diagnostics/ProfileManager.cs [hunk 10]
```diff
@@ -18,16 +18,24 @@ internal static class ProfileManager
     {
         private const int ProcessExitTimeoutInSeconds = 180;
 
+        private const string ServiceProfilerAgentGuid = "F5091AA9-80DC-49FF-A7CF-BD1103FE149D";
+        private const string ServiceProfilerProviderGuid = "1C92BA2A-A990-480F-A02F-40068871CAAC";
```
Comment (pharring) [inline Kudu.Services/Diagnostics/ProfileManager.cs]: Let's call this ```DetailedTracingAgentGuid```

File: Kudu.Services/Diagnostics/ProfileManager.cs [hunk 11]
```diff
@@ -39,34 +47,69 @@ internal static class ProfileManager
...
 
-                EnsureIdleTimer();
+                EnsureIdleTimer(iisProfiling);
 
                 return new ProfileResultInfo(HttpStatusCode.OK, string.Empty);
             }
         }
 
-        internal static async Task<ProfileResultInfo> StopProfileAsync(int processId, ITracer tracer = null)
+        private static async Task<ProfileResultInfo> StartIISSessionAsync(int processId, int profilingSessionId, ITracer tracer = null)
```
Comment (pharring) [inline Kudu.Services/Diagnostics/ProfileManager.cs]: How about some comments explaining what's going on here?
e.g. Start a session with the "custom etw provider agent", add in the "detailed tracing agent", turn on the IISWWWServer provider.

File: Kudu.Services.Web/Content/Scripts/ProcessExplorer.js [hunk 12]
```diff
@@ -374,13 +397,19 @@ var Process = (function () {
...
         }
 
+        var iisProfilingCheckbox = Utilities.getCheckbox(this._json.id + "-iisProfilingCheck", "Collect IIS Events")
+       
         if (Process.prototype.WebSiteSku === "Free" || Process.prototype.WebSiteSku === "Shared") {
             $(profilingButton).css("opacity", "0.5");
             $(profilingButton).off("click");
             $(profilingButton).attr("title", "Profiling is not supported for Free/Shared sites.");
             $(profilingButton).tooltip().show();
+
```
Comment (safihamid) [inline Kudu.Services.Web/Content/Scripts/ProcessExplorer.js]: nit: extra line

File: Kudu.Services.Web/Content/Scripts/ProcessExplorer.js [hunk 13]
```diff
@@ -920,26 +949,36 @@ function handleProfilingEvents(e, processId) {
         return;
     }
 
+    var iisProfiling = false;
+
+    if (document.getElementById(processId + "-iisProfilingCheck") !=null) {
```
Comment (safihamid) [inline Kudu.Services.Web/Content/Scripts/ProcessExplorer.js]: nit: need space, != null

File: Kudu.Services.Web/Content/Scripts/ProcessExplorer.js [hunk 14]
```diff
@@ -985,6 +1024,7 @@ function stopProfiling(e, processId) {
 
     Utilities.downloadURL(uri, true);
     e.target.textContent = "Start Profiling";
+
```
Comment (safihamid) [inline Kudu.Services.Web/Content/Scripts/ProcessExplorer.js]: nit: extra line.

File: Kudu.Services/Diagnostics/ProfileManager.cs [hunk 15]
```diff
@@ -18,16 +18,24 @@ internal static class ProfileManager
...
+        private const string IISWWWServerProviderGuid = "3A2A4E84-4C21-4981-AE10-3FDA0D9B0F83";
+        private const string DiagnosticsHubAgentGuid = "4EA90761-2248-496C-B854-3C0399A591A4";
+
         private static ConcurrentDictionary<int, ProfileInfo> _profilingList = new ConcurrentDictionary<int, ProfileInfo>();
         private static object _lockObject = new object();
 
         // The profiling session timeout, this is temp fix before VS2015 Update 1.
         private static readonly TimeSpan _profilingTimeout = TimeSpan.FromMinutes(15);
+
+        private static readonly TimeSpan _profilingIISTimeout = TimeSpan.FromMinutes(3);
```
Comment (safihamid) [inline Kudu.Services/Diagnostics/ProfileManager.cs]: nit: _profilingIisTimeout

File: Kudu.Services/Diagnostics/ProfileManager.cs [hunk 16]
```diff
@@ -18,16 +18,24 @@ internal static class ProfileManager
     {
         private const int ProcessExitTimeoutInSeconds = 180;
 
+        private const string UserModeCustomProviderAgentGuid = "F5091AA9-80DC-49FF-A7CF-BD1103FE149D";
+        private const string DetailedTracingAgentGuid = "1C92BA2A-A990-480F-A02F-40068871CAAC";
+        private const string IISWWWServerProviderGuid = "3A2A4E84-4C21-4981-AE10-3FDA0D9B0F83";
```
Comment (safihamid) [inline Kudu.Services/Diagnostics/ProfileManager.cs]: nit: use standard pascal casing for acronyms longer than 2 characters. this better be called IisWebServerProviderGuid

File: Kudu.Services/Diagnostics/ProfileManager.cs [hunk 17]
```diff
@@ -39,34 +47,71 @@ internal static class ProfileManager
...
-                
-                string arguments = System.Environment.ExpandEnvironmentVariables(string.Format("start {0} /attach:{1} /loadAgent:4EA90761-2248-496C-B854-3C0399A591A4;DiagnosticsHub.CpuAgent.dll  /scratchLocation:%LOCAL_EXPANDED%\\Temp", profilingSessionId, processId));
 
-                var profileProcessResponse = await ExecuteProfilingCommandAsync(arguments, tracer);
+                if (iisProfiling)
+                {
+                    var profileResponseStatus = await StartIISSessionAsync(processId, profilingSessionId, tracer);
 
-                if (profileProcessResponse.StatusCode != HttpStatusCode.OK)
+                    if (profileResponseStatus.StatusCode != HttpStatusCode.OK)
```
Comment (safihamid) [inline Kudu.Services/Diagnostics/ProfileManager.cs]: This logic is repeated in both if and else part, move it outside so it is not repeated.

File: Kudu.Services/Diagnostics/ProfileManager.cs [hunk 18]
```diff
@@ -39,34 +47,71 @@ internal static class ProfileManager
...
 
-                EnsureIdleTimer();
+                EnsureIdleTimer(iisProfiling);
 
                 return new ProfileResultInfo(HttpStatusCode.OK, string.Empty);
             }
         }
 
-        internal static async Task<ProfileResultInfo> StopProfileAsync(int processId, ITracer tracer = null)
+        private static async Task<ProfileResultInfo> StartIISSessionAsync(int processId, int profilingSessionId, ITracer tracer = null)
```
Comment (safihamid) [inline Kudu.Services/Diagnostics/ProfileManager.cs]: rename the method to StartIisSessionAsync

File: Kudu.Services/Diagnostics/ProfileManager.cs [hunk 19]
```diff
@@ -83,30 +128,58 @@ internal static class ProfileManager
...
+                profileFileName = string.Format("profile_{0}_{1}_{2}.diagsession", InstanceIdUtility.GetShortInstanceId(), System.Diagnostics.Process.GetProcessById(processId).ProcessName, processId);
+            }
             return System.Environment.ExpandEnvironmentVariables("%LOCAL_EXPANDED%\\Temp\\" + profileFileName);
         }
 
         internal static bool IsProfileRunning(int processId)
         {
             return _profilingList.ContainsKey(processId);
         }
+        internal static bool IsIISProfileRunning(int processId)
```
Comment (safihamid) [inline Kudu.Services/Diagnostics/ProfileManager.cs]: nit: rename to IsIisProfilingRunning

File: Kudu.Services/Diagnostics/ProfileManager.cs [hunk 20]
```diff
@@ -83,30 +128,58 @@ internal static class ProfileManager
...
+        {
+            bool iisProfiling = false;
+            if (_profilingList != null)
+            {
+                ProfileInfo profilingId;
+                _profilingList.TryGetValue(processId, out profilingId);
+
+                if (profilingId != null)
+                    iisProfiling = profilingId.IsIISProfiling;
+
```
Comment (safihamid) [inline Kudu.Services/Diagnostics/ProfileManager.cs]: nit: extra line

Comment (safihamid) [inline Kudu.Services/Diagnostics/ProfileManager.cs]: nit: the one liner needs still be in curly brackets. 

if (profilingId != null)
{
          iisProfiling = profilingId.IsIISProfiling;
}

File: Kudu.Services/Diagnostics/ProfileManager.cs [hunk 21]
```diff
@@ -83,30 +128,58 @@ internal static class ProfileManager
...
+            }
+
+            return iisProfiling;
+        }
+
```
Comment (safihamid) [inline Kudu.Services/Diagnostics/ProfileManager.cs]: nit: please remove all these extra lines

File: Kudu.Services/Diagnostics/ProfileManager.cs [hunk 22]
```diff
@@ -295,15 +386,19 @@ public ProfileResultInfo(HttpStatusCode statusCode, string message)
...
+                this.IsIISProfiling = iisProfiling;
             }
 
             public int SessionId { get; set; }
-
+           
             public DateTime StartTime { get; set; }
+
+            public bool IsIISProfiling { get; set; }
```
Comment (safihamid) [inline Kudu.Services/Diagnostics/ProfileManager.cs]: nit: IsIisProfiling

**Comment by ahmelsayed** [comment]
What would be the negatives of always including the IIS providers in the profile?

**Comment by safihamid** [comment]
@ahmelsayed IIS profiles use /threadTime which log all context switches, block times, ... creating such profiles are very CPU intensive and generate large files. we definitely don;' want to enable them when running CPU profiling which just gets a CPU call stack every millisecond and can be opened in VS for further processing. IIS profiles are for deeper investigations and can be opened in perfView.

**Comment by safihamid** [comment]
Can you deploy a standard asp.net site or web api, on a small dedicated worker, enable IIS profiling and load test it with WCAT or some other http stress testing tool for 3 minutes and share the file? Also plz log the timestamp, I want to see how big the file would be and how the VM behaved during that time.
If the file size is too big that even PerfView has problem opening it or if the VM is having Disk Space issues on D drive or the site's temp folder then we must lower the timeout value

File: Kudu.Services.Web/Content/Scripts/ProcessExplorer.js [hunk 23]
```diff
@@ -374,13 +397,18 @@ var Process = (function () {
             $(profilingButton).addClass("btn-danger");
         }
 
+        var iisProfilingCheckbox = Utilities.getCheckbox(this._json.id + "-iisProfilingCheck", "Collect IIS Events")
+       
         if (Process.prototype.WebSiteSku === "Free" || Process.prototype.WebSiteSku === "Shared") {
```
Comment (safihamid) [inline Kudu.Services.Web/Content/Scripts/ProcessExplorer.js]: The profiling checkbox should also be disabled(not clickable) for free/shared. Similar to what we are doing for the button.

File: Kudu.Services.Web/Content/Scripts/ProcessExplorer.js [hunk 24]
```diff
@@ -920,26 +948,36 @@ function handleProfilingEvents(e, processId) {
         return;
     }
 
+    var iisProfiling = false;
```
Comment (safihamid) [inline Kudu.Services.Web/Content/Scripts/ProcessExplorer.js]: As we discussed before, we need to show a message to the user when they check the iis profiling checkbox. The message should have 'ok' and 'cancel' buttons and proceed if 'ok' is clicked.
The message should specify what IIS profiling does, and the fact that 'This is a relatively expensive option' and the file needs to be opened in 'PerfView' with a link to where to download it from. Should also most likely specift the 3 minute timeout or whatever timeout we pick.

Comment (puneet-gupta) [inline Kudu.Services.Web/Content/Scripts/ProcessExplorer.js]: Will do

File: Kudu.Services.Web/Content/Scripts/ProcessExplorer.js (hunk 23 above)
Comment (puneet-gupta) [inline Kudu.Services.Web/Content/Scripts/ProcessExplorer.js]: Will do

**Comment by puneet-gupta** [comment]
Working on the stress test and repro part....should have it ready in a day or two

**Comment by puneet-gupta** [comment]
@safihamid - Instead of hardcoding the _profileIisTimeout to 3 minutes (or 1 minute if we end up finalizing it), are we ok to expose an APP_SETTING something like IIS_PROFILING_TIMEOUT_KUDU which an end customer can site and that can override the default timeout ??? 

In some situations, I am thinking that there may be a valid need to increase this timeout and we should have some option for the customer. 

We can always have the MAX to be 15 minutes so that if a customer chooses a higher value than this, it doesn't get honored.

**Comment by safihamid** [comment]
@puneet-gupta  That's actually a good idea. we can have the default and then have an option to override if really needed. Also I don't remember if I have ever tested it but what happens to the UI if profiling is started and then they forget to stop it and it times out?

**Comment by puneet-gupta** [comment]
@safihamid "Also I don't remember if I have ever tested it but what happens to the UI if profiling is started and then they forget to stop it and it times out?"    

I tested this and found that the UI doesn't auto-update itself. However if someone clicks REFRESH or Refreshes the browser after the timeout then the button status is correctly set to Start Profiling

Shall I also implement the logic for IIS_PROFILING_TIMEOUT_KUDU  and does this name look OK?

**Comment by safihamid** [comment]
@puneet-gupta how about  IIS_PROFILING_TIMEOUT_OVERRIDE_IN_SECONDS?

**Comment by puneet-gupta** [comment]
seems to long, how about **IIS_PROFILING_TIMEOUT_IN_SECONDS**

**Comment by safihamid** [comment]
good! :-)

**Comment by puneet-gupta** [comment]
cool , will update this tomorrow

**Comment by puneet-gupta** [comment]
@safihamid also shared a sample ETW trace with you over email to have a quick review

File: Kudu.Services/Diagnostics/ProfileManager.cs [hunk 25]
```diff
@@ -35,6 +35,16 @@ internal static class ProfileManager
 
         private static string _processName = System.Environment.ExpandEnvironmentVariables("%SystemDrive%\\msvsmon\\profiler\\VSStandardCollector.Dev14.exe");
 
+        static ProfileManager()
+        {
```
Comment (safihamid) [inline Kudu.Services/Diagnostics/ProfileManager.cs]: Please combine this logic with private static readonly TimeSpan _profilingIisTimeout = TimeSpan.FromMinutes(3);
and remove this static assignment.

File: Kudu.Services/Diagnostics/ProfileManager.cs [hunk 26]
```diff
@@ -104,6 +116,26 @@ internal static class ProfileManager
...
+            int timeout = -1;
+
+            string iisProfilingTimeoutInSeconds = System.Environment.GetEnvironmentVariable("APPSETTING_IIS_PROFILING_TIMEOUT_IN_SECONDS");
+
+            if (!string.IsNullOrEmpty(iisProfilingTimeoutInSeconds))
+            {
+                if (Int32.TryParse(iisProfilingTimeoutInSeconds, out timeout))
+                {
+                    if (timeout>(60*15)) // making sure that no configures it more than 15 minutes
```
Comment (safihamid) [inline Kudu.Services/Diagnostics/ProfileManager.cs]: Instead of 60*15 please use 60*_profilingTimeout, also fix the spacing and comment

File: Kudu.Contracts/Diagnostics/ProcessInfo.cs [hunk 27]
```diff
@@ -30,6 +30,12 @@ public class ProcessInfo : INamedObject
         [JsonProperty(PropertyName = "is_profile_running", DefaultValueHandling = DefaultValueHandling.Ignore)]
         public bool IsProfileRunning { get; set; }
 
+        [JsonProperty(PropertyName = "is_iis_profile_running", DefaultValueHandling = DefaultValueHandling.Ignore)]
+        public bool IsIisProfileRunning { get; set; }
+
+        [JsonProperty(PropertyName = "iis_profile_timeout_seconds", DefaultValueHandling = DefaultValueHandling.Ignore)]
```
Comment (safihamid) [inline Kudu.Contracts/Diagnostics/ProcessInfo.cs]: I would say let's call this: iis_profile_timeout_in_seconds to be compat with rest of the code

File: Kudu.Services.Web/Content/Scripts/ProcessExplorer.js [hunk 28]
```diff
@@ -943,23 +960,73 @@ function searchForHandle() {
...
+                divConfirm = Utilities.createDiv('dialog-confirm');
+                divConfirm.title = "Collect IIS Events?";
+
+            }
+
+            divConfirm.innerHTML = "IIS profiling enables IIS and threadtime ETW events. The generated trace file can be analyzed using Perview which is available at <a style='outline: none' href='https://www.microsoft.com/en-us/download/details.aspx?id=28567'>https://www.microsoft.com/en-us/download/details.aspx?id=28567</a>. <br/><br/>The profiling session will automatically timeout after <b>" + iisProfilingTimeoutInSeconds.toString() + "</b> seconds if not stopped manually.<br/><br/>Enabling IIS Profiling is a relatively <b>expensive option</b> to turn on as it increases the CPU usage and disk I/O on your instance. Are you sure you want to continue ?"
+
+            $(divConfirm).dialog({
+                resizable: false,
+                height: 330,
```
Comment (safihamid) [inline Kudu.Services.Web/Content/Scripts/ProcessExplorer.js]: Nit: not a javascript expert but are we supposed to hard-code the sizes here? how does it show on different resolutions of hard coded? maybe check rest of Kudu, if they are same then ok.

File: Kudu.Services/Diagnostics/ProfileManager.cs (hunk 25 above)
Comment (puneet-gupta) [inline Kudu.Services/Diagnostics/ProfileManager.cs]: Sorry not sure if I am understanding properly, you want me to remove this line 

 `private static readonly TimeSpan _profilingIisTimeout = TimeSpan.FromMinutes(3);`

or the static constructor ?
```
static ProfileManager()
        {
            int iisProfilingTimeout = GetIisProfilingTimeout();

            if (iisProfilingTimeout > 0)
            {
                _profilingIisTimeout = TimeSpan.FromSeconds(iisProfilingTimeout);
            }
        }
```

File: Kudu.Services/Diagnostics/ProfileManager.cs (hunk 26 above)
Comment (puneet-gupta) [inline Kudu.Services/Diagnostics/ProfileManager.cs]: not following...This is not actually a multiple of _profliingTimeout. This is an upper limit of _profilingTimeout....Looks like your comment is missing some characters...

Comment (safihamid) [inline Kudu.Services/Diagnostics/ProfileManager.cs]: @puneet-gupta _profilingTimeout is already set to 15 minutes. So I am suggesting to use that value instead. so replcae 60 x 15 with 60 x _profilingTimeout

File: Kudu.Services/Diagnostics/ProfileManager.cs (hunk 25 above)
Comment (safihamid) [inline Kudu.Services/Diagnostics/ProfileManager.cs]: @puneet-gupta plz remove the assignment here: private static readonly TimeSpan _profilingIisTimeout = TimeSpan.FromMinutes(3);
and add that logic to the static constructor.

File: Kudu.Services.Web/Content/Scripts/ProcessExplorer.js (hunk 24 above)
Comment (puneet-gupta) [inline Kudu.Services.Web/Content/Scripts/ProcessExplorer.js]: This is done now !!!

**Comment by puneet-gupta** [comment]
@safihamid  Regarding the specifying height and width, I do see hard-coded height and width at some places in code. If I am leaving this to AUTO then the dialog is looking a bit ugly as there is quite some information displayed in the message box.

It seems I have addressed the issues but please let me know if I missed something (It seems some comments are missing for some reason but I remember them reading here)

File: Kudu.Services/Diagnostics/ProfileManager.cs [hunk 29]
```diff
@@ -18,16 +18,29 @@ internal static class ProfileManager
...
+        private const string IisWebServerProviderGuid = "3A2A4E84-4C21-4981-AE10-3FDA0D9B0F83";
+        private const string DiagnosticsHubAgentGuid = "4EA90761-2248-496C-B854-3C0399A591A4";
+        
         private static ConcurrentDictionary<int, ProfileInfo> _profilingList = new ConcurrentDictionary<int, ProfileInfo>();
         private static object _lockObject = new object();
 
         // The profiling session timeout, this is temp fix before VS2015 Update 1.
         private static readonly TimeSpan _profilingTimeout = TimeSpan.FromMinutes(15);
+
+        private static readonly TimeSpan _profilingIisTimeout = TimeSpan.MinValue;
```
Comment (pharring) [inline Kudu.Services/Diagnostics/ProfileManager.cs]: You can write
```C#
private static readonly TimeSpan _profilingIisTimeout = GetIisProfilingTimeout();
```

and then eliminate the static constructor entirely.

Comment (puneet-gupta) [inline Kudu.Services/Diagnostics/ProfileManager.cs]: Thanks Paul, checked in this change !!!

**Comment by davidebbo** [comment]
I merged the change. Thanks @puneet-gupta for your contrib!

**Comment by puneet-gupta** [comment]
Thank you Hamid , Paul and David for letting me contribute to this project !!!! It was a great experience for me and I am sure it will help support and customers tremendously....

Do we know when this available to all in Kudu ?

**Comment by davidebbo** [comment]
Hopefully we'll deployit next week.
